test/StateDiffChecker.t.sol

```diff
@@ -0,0 +1,224 @@
+        expectedDiff_ = Checker.parseDiffSpecs(json);
+
+        actualDiff_ = Checker.extractDiffSpecFromAccountAccesses(accountAccesses);
+    }
+
+    /// @dev Test that the state diff generated by running Target.set() matches the expected state diff
+    ///      as defined in testDiff.json
+    function test_checkStateDiff_succeeds() public {
+        (Checker.StateDiffSpec memory expectedDiff, Checker.StateDiffSpec memory actualDiff) = _executeAndGetDiffs();
+        Checker.checkStateDiff(expectedDiff, actualDiff);
+    }
+
+    /// @dev Test that the correct error is thrown when the chain ID does not match.
+    function test_checkStateDiff_chainIdMismatch_reverts() public {
+        (Checker.StateDiffSpec memory expectedDiff, Checker.StateDiffSpec memory actualDiff) = _executeAndGetDiffs();
+        actualDiff.chainId = 31338;
+        vm.expectRevert(abi.encodeWithSelector(StateDiffMismatch.selector, "chainId", 31337, 31338));
+        Checker.checkStateDiff(expectedDiff, actualDiff);
+    }
+
+    /// @dev Test that the correct error is thrown when the number of storage modifications does not match.
+    function test_checkStateDiff_lengthMismatch_reverts() public {
+        (Checker.StateDiffSpec memory expectedDiff,) = _executeAndGetDiffs();
+        Checker.StateDiffSpec memory shorterDiff = Checker.StateDiffSpec({
+            chainId: 31337,
+            storageSpecs: new Checker.StorageDiffSpec[](expectedDiff.storageSpecs.length - 1)
+        });
+        shorterDiff.storageSpecs[0] = expectedDiff.storageSpecs[0];
+        vm.expectRevert(
+            abi.encodeWithSelector(
+                StateDiffMismatch.selector,
+                "storageSpecs.length",
+                expectedDiff.storageSpecs.length,
+                shorterDiff.storageSpecs.length
+            )
+        );
+        Checker.checkStateDiff(expectedDiff, shorterDiff);
+    }
+
+    /// @dev A utility function to copy a StateDiffSpec struct so that we can modify it without affecting the original.
+    function copyDiff(Checker.StateDiffSpec memory diff) internal pure returns (Checker.StateDiffSpec memory) {
+        Checker.StateDiffSpec memory copy;
+        copy.chainId = diff.chainId;
+        copy.storageSpecs = new Checker.StorageDiffSpec[](diff.storageSpecs.length);
+        for (uint256 i = 0; i < diff.storageSpecs.length; i++) {
+            copy.storageSpecs[i].account = diff.storageSpecs[i].account;
+            copy.storageSpecs[i].slot = diff.storageSpecs[i].slot;
+            copy.storageSpecs[i].newValue = diff.storageSpecs[i].newValue;
+            copy.storageSpecs[i].previousValue = diff.storageSpecs[i].previousValue;
+
+            // Ensure that all values have been copied over
+            require(
+                keccak256(abi.encode(copy.storageSpecs[i])) == keccak256(abi.encode(diff.storageSpecs[i])),
+                "StateDiffCheckerTest: StorageDiffSpec copying failed"
+            );
```

Comment:
Probably better to use `keccak256(abi.encode(copy)) == keccak256(abi.encode(diff))` outside of the loop to handle if any top-level fields were added